Document Ghc 69925 (#95)
haskell/error-messages: message-index/messages/GHC-69925/index.md

---
title: Illegal unboxed string literal in pattern
summary: Illegal unboxed string literal in pattern
severity: error
introduced: 9.6.1
---

> When pattern matching, unboxed string literals (with the `MagicHash` postfix) are not allowed. `"foo"#` is of type `Addr#` -- that is, "An arbitrary machine address assumed to point outside the garbage-collected heap.".

Reviewer: ```suggestion
When pattern matching, unboxed string literals (with the `MagicHash` postfix) are not allowed. Unboxed string literals are essentially C strings allocated outside of the Haskell heap, and they evaluate to pointers to the resulting strings. Thus,`"foo"#` is of type `Addr#`. While `eqAddr#` can be used to compare these pointers, two identically-written unboxed string literals typically do not point the same address.".
```

Reviewer: I think the fix in the example should be to use boxed string literals instead :-)

> When pattern matching, unboxed string literals (with the `MagicHash` postfix) are not allowed. Unboxed string literals are essentially C strings allocated outside of the Haskell heap, and they evaluate to pointers to the resulting strings. Thus,`"foo"#` is of type `Addr#`. While `eqAddr#` can be used to compare these pointers, two identically-written unboxed string literals typically do not point the same address.".

Reviewer: ```suggestion
When pattern matching, unboxed string literals (with the `MagicHash` postfix) are not allowed. Unboxed string literals are essentially C strings allocated outside of the Haskell heap, and they evaluate to pointers to the resulting strings. Thus,`"foo"#` is of type `Addr#`. While `eqAddr#` can be used to compare these pointers, two identically-written unboxed string literals typically do not point the same address.
```

> Instead, consider using the unboxed function that implements equality (`eqAddr#`) which can then be used in combination with `if .. then .. else` via `isTrue#`.

Reviewer: ```suggestion
```

---
title: Illegal unboxed string literal in pattern
summary: Illegal unboxed string literal in pattern
severity: error
introduced: 9.6.1
---

> When pattern matching, unboxed string literals (with the `MagicHash` postfix) are not allowed. Unboxed string literals are essentially C strings allocated outside of the Haskell heap, and they evaluate to pointers to the resulting strings. Thus,`"foo"#` is of type `Addr#`. While `eqAddr#` can be used to compare these pointers, two identically-written unboxed string literals typically do not point the same address.

Reviewer 2: How about cross-referencing docs for magic hash?

```suggestion
When pattern matching, unboxed string literals (with the [MagicHash](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/exts/magic_hash.html?highlight=magichash
) postfix) are not allowed. Unboxed string literals are essentially C strings allocated outside of the Haskell heap, and they evaluate to pointers to the resulting strings. Thus,`"foo"#` is of type `Addr#`. While `eqAddr#` can be used to compare these pointers, two identically-written unboxed string literals typically do not point the same address.
```
